#167128184 Users can view and update their personal information

MandelaK opened the pull request:
## Description ##
- Allow users to view and update their profile. [You can test it live here ](https://landville-frontend-pr-12.herokuapp.com/profile)

## Type of change ##
Please select the relevant option
- [ ] Bug fix(a non-breaking change which fixes an issue)
- [x] New feature(a non-breaking change which adds functionality)
- [ ] Breaking change(fix of a feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested? ##
Please describe the tests that you ran to verify your changes
- [ ] End to end test
- [x] Integration test
- [x] unit test

## Checklist: ##
- [x] My code follows the style guidelines of this project
- [x] I have linted my code prior to submission
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

## Related Pivotal Tracker stories ##
[#167128184](https://www.pivotaltracker.com/story/show/167128184)

MandelaK committed:
ch(storybook) Set up storybook

- Configure storybook for angular application

[Finishes #167205870]

MandelaK committed:
ch(SCSS) Configure SCSS and set up project folders

- Configures SCSS support for the application
- Set up folder structure with placeholder files
- Add image assets

[Finishes #167205870]

MandelaK committed:
ft(profile) Users can view and update their profile

- Users can view and update their profile

[Starts #167128184]

MandelaK committed:
ft(services) set up services

- set up services for profile component

[#167128184]

MandelaK committed:
ft(profile view) Improve styling and improve coverage

- Add more CSS styling to match mockup
- Add new services for reusablity

[#167128184]

MandelaK committed:
ft(profile) Refactor code and improve coverage

- Improve the code coverage
- Refactor profile form to use Reactive Forms instead of templates
- Do away with the profile interceptor, custom phone directives
- Fix issues with inconsistent test results

[#167128184]

MandelaK committed:
ft(profile): Fix inconsitencies with failing tests

- Fix issues with failing tests due to poorly setup mock files

[#167128184]

kelvinndmo commented on src/app/pages/profile/profile-sidebar/profile-sidebar.component.scss:
```diff
@@ -0,0 +1,112 @@
+  }
+  &-image {
+    &-container {
+      // height: 250px;
```
remove this commented code here and any other place

MandelaK replied:
I fixed this

kelvinndmo commented on src/app/shared/services/profile/profile.service.ts:
```diff
@@ -0,0 +1,48 @@
+  providedIn: 'root'
+})
+export class ProfileService {
+  profileUrl = 'http://127.0.0.1:8000/api/v1/auth/profile/';
```
I believe this should not be here

MandelaK replied:
Hi, do you have a suggestion of how this can be fixed? When I move the URL to the environment file, I keep getting an error ... not sure how to go about this

kelvinndmo replied:
hey did you fix this @MandelaK

MandelaK replied:
I think so ...  like I said earlier, I don't really understand the change you wanted me to make, maybe you can comment on that?

kelvinndmo replied:
the profile url i meant should be in the config, but i just saw you did so, good work on this.

MandelaK committed:
ft(implement feedback) Implement feedback and refactor code

- Add custom styling for profile form button
- Remove hover and transition effects from profile form
- Update FontAwesome CDN version

[#167128184]

MandelaK committed:
ft(profile) Refactor code and improve coverage

- Use subjects to improve how data is shared between components
- Improve coverage
- Adopt folder structure on develop

[Finishes #167128184]

kelvinndmo commented on src/app/helpers/spies.ts:
```diff
@@ -1,21 +1,31 @@
-      });
+  spies.forEach(spyObj => {
+    Object.keys(spyObj).forEach(prop => {
+      if (typeof spyObj[prop] === 'function') {
```
there is no else for this statement which beats the logic of it being a conditional statement, so i guess you can as well remove the if and just point to what is returned by the if

MandelaK replied:
But I only want to reset spy functions ... if the object is not a function we do nothing to it ... If I remove the `statement` then some unwanted bugs might be introduced ... regardless, this statement wasn't authored by me, it just says so because the linter made some changes, but if you are keen you'll notice that it was introduced by a previous PR

kelvinndmo replied:
ooh...ok..no big deal as it does not break anything

CryceTruly commented on src/app/pages/profile/profile-sidebar/profile-sidebar.component.html:
```diff
@@ -0,0 +1,60 @@
+  <div class="profile-image-container">
+    <ngx-spinner></ngx-spinner>
+
+    <img
```
@MandelaK Good work man, However it can be really efficient if users can not upload media files that are not images. We can achieve this by adding an `accept` attribute to the img tag.

MandelaK replied:
Thanks for pointing this out, I have fixed it

MandelaK committed:
ft(responsiveness) Improve profile responsiveness

- Set media queries to achieve responsiveness
- Users can only select images when they want to upload an image
- Add custom styling rules for Mozilla browser

[Finishes #167128184]

patrickf949 reviewed (approved):
Responsive enough - 👍🏿
Follows Conventions - 👍🏿
Neat implementation - 👍🏿
LGTM - definitely

JamesMudidi commented on src/app/pages/profile/personal-information/personal-information.component.ts:
```diff
@@ -0,0 +1,171 @@
+      City: this.profileForm.controls.city.value,
+      State: this.profileForm.controls.state.value
+    };
+    // the backend expects to pick address from an `address` dictionary that
```
Nice work @MandelaK this is really impressive but I think comments on Javascript code takes another approach which uses /* */ instead of //

MandelaK replied:
This is single-line comment and so I just use the double-slash ... /* is meant for blocks of comments, which is not the case here

JamesMudidi reviewed (changes requested):
The work you have done is incredible but I noticed that the app is not as responsive as it should be when tested against the different screen sizes as demonstrated in the .gif file below. I also noticed that the profile image gets squashed when in tablet mode.

![Jul-31-2019 10-56-47 PM](https://user-images.githubusercontent.com/39102955/62243842-ccb86f80-b3e6-11e9-81da-abebbdf0f3d7.gif)

Victoradukwu reviewed (approved):
LGTM

MandelaK committed:
ft(feedback) Improve responsiveness

- Improve responsiveness
- Hide placeholder so that it is only displayed when clicked
- Rollback upgdate to FontAwesome 4.7
- Fix broken links on header and footer icons

[Finishes #167128184]

JamesMudidi reviewed (approved):
LGTM

akram256 reviewed (approved):
LGTM

bl4ck4ndbr0wn reviewed (approved):
great work. LGTM

MandelaK committed:
ft(UI) Update navbar

- Update navbar to improve UX/UI

[Finishes #167128184]
